src/xcode/ENA/ENA/Source/Views/BaseElements/ENATanInput.swift

```diff
@@ -104,27 +130,98 @@ class ENATanInput: UIControl, UIKeyInput {
 	}
 
 	func insertText(_ text: String) {
+		guard !inputBlocked else { return }
 		for character in text {
 			guard !isValid else { return }
-			let label = stackView.arrangedSubviews[count] as? UILabel
-			label?.text = "\(character)"
-			self.text += "\(character)"
+			let label = labels[count]
+
+			label.text = "\(character.uppercased())"
+			self.text += "\(character.uppercased())"
+			if let enaInputLabel = label as? ENATanInputLabel {
+				if forbiddenArray.contains(String(character.uppercased())) {
+					enaInputLabel.isValid = false
+					enaInputLabel.textColor = .enaColor(for: .textSemanticRed)
+					inputBlocked = true
+				} else {
+					enaInputLabel.isValid = true
+					enaInputLabel.textColor = labelTextColor
+				}
+			}
+
 		}
-		delegate?.tanChanged(isValid: isValid)
+		delegate?.tanChanged(isValid: isValid, checksumIsValid: verifyChecksum(input: self.text), isBlocked: inputBlocked
+		)
 	}
 
 	func deleteBackward() {
 		guard !isEmpty else { return }
+		inputBlocked = false
 		text = String(text[..<text.index(before: text.endIndex)])
-		let label = stackView.arrangedSubviews[count] as? UILabel
-		label?.text = ""
-		delegate?.tanChanged(isValid: isValid)
+		let label = labels[count]
+		label.text = ""
+		if let enaInputLabel = label as? ENATanInputLabel {
+			enaInputLabel.isValid = true
+		}
+		delegate?.tanChanged(isValid: isValid, checksumIsValid: false, isBlocked: inputBlocked)
 	}
 
 	func clear() {
-		for i in 0 ..< text.count {
-			(stackView.arrangedSubviews[i] as? UILabel)?.text = ""
-		}
+		labels.forEach { $0.text = "" }
 		text = ""
 	}
+	
+	func verifyChecksum(input: String) -> Bool {
+		guard isValid else {
+			return false
+
+		}
+		let start = input.index(input.startIndex, offsetBy: 0)
+		let end = input.index(input.startIndex, offsetBy: input.count - 2)
+		let testString = String(input[start...end])
+		return input.last == calculateChecksum(input: testString)
+	}
+	
+	func calculateChecksum(input: String) -> Character {
```

Comment:
Semantically this should be handled in the corresponding controller. But is fine for now.